### PR TITLE
Fix how discounts are displayed in cart summary when using vouchers

### LINKED ISSUE #9702: [BOOM-5101] Voucher free shipping show 0.00% instead of free sipping
## SPECS

Cart rule with free shipping only: Display free shipping instead of 0,00%
Cart rule with free shipping + discount (amount or %): Display price of the carrier + Discount amount added
Cart rule with % discount: Display the discount amount            

image

You should always see the same value between the "discount(s)" line and the "cart rule" line if there is only one cart rule in the cart

----------------------
> This issue has been migrated from the Forge. Read the [original ticket here](http://forge.prestashop.com/browse/BOOM-5101).

- _**Reporter:**_ PN
- _**Created at:**_ Thu, 15 Mar 2018 09:27:45 +0100

<p>Hello,</p>

<p> </p>

<p><b>Type of view discounts in cart when is discount action set as free shipping. Why is showed: -0.00%? Better will be show free shipping instead of -0.00%.</b></p>

<p>Thank you</p>

<p>Best regards</p>

<p>PN</p>
        
* How to reproduce the issue ?

Catalog, Discounts, Cart Rules
<br/>

<br/>
for:
<br/>
1) Set Highlight: YES
<br/>
2) Leave field code blank
<br/>
3) Set actions: Free shipping: YES, none

### PR DESCRIPTION
**Pull request by matthieu-rolland**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | See below for display rules
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #9702
| How to test?  | See below

On the cart summary, where reductions are displayed, here are the rules that should be respected:

- Cart rule with free shipping only: Display free shipping instead of 0,00%  
- Cart rule with free shipping + discount (amount or %): Display price of the carrier + Discount amount added  
- Cart rule with % discount: Display the discount amount (not the percentage)

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/16235)

### CODE REVIEW

File: src/Adapter/Presenter/Cart/CartPresenter.php [hunk 1]
```diff
@@ -585,6 +608,44 @@ private function getTemplateVarVouchers(Cart $cart)
...
+    /**
+     * @param array $cartVoucher
+     *
+     * @return bool
+     */
+    private function cartVoucherHasAmountReduction($cartVoucher)
+    {
+        return isset($cartVoucher['reduction_amount']) && $cartVoucher['reduction_amount'] > 0;
+    }
+
```
Comment (matthieu-rolland) [inline src/Adapter/Presenter/Cart/CartPresenter.php:642]: Those are three legacy conditions that I extracted in functions with a self explaining name for readability purpose.

File: src/Adapter/Presenter/Cart/CartPresenter.php [hunk 2]
```diff
@@ -546,28 +546,51 @@ private function getTemplateVarVouchers(Cart $cart)
...
                     }
                 }
-                // following will do currency conversion to current one
-                $cartVoucher['reduction_formatted'] = $this->priceFormatter->convertAndFormat($value);
+                $freeShippingOnly = false;
+            }
+            // when a voucher has only a shipping reduction, the value displayed must be "Free Shipping"
+            if ($freeShippingOnly) {
+                $cartVoucher['reduction_formatted'] = $this->translator->trans(
+                    'Free Shipping',
```
Comment (LouiseBonnard) [inline src/Adapter/Presenter/Cart/CartPresenter.php]:
```suggestion
                    'Free shipping',
```

File: src/Adapter/Presenter/Cart/CartPresenter.php [hunk 3]
```diff
@@ -546,28 +546,51 @@ private function getTemplateVarVouchers(Cart $cart)
...
+                    [],
+                    'Admin.Orderscustomers.Feature'
```
Comment (LouiseBonnard) [inline src/Adapter/Presenter/Cart/CartPresenter.php]:
```suggestion
                    'Admin.Shipping.Feature'
```

File: src/Adapter/Presenter/Cart/CartPresenter.php (hunk 2 above)
Comment (matthieu-rolland) [reply src/Adapter/Presenter/Cart/CartPresenter.php]: done 👍

File: src/Adapter/Presenter/Cart/CartPresenter.php (hunk 3 above)
Comment (matthieu-rolland) [reply src/Adapter/Presenter/Cart/CartPresenter.php]: done 👍

File: src/Adapter/Presenter/Cart/CartPresenter.php [hunk 4]
```diff
@@ -585,6 +608,44 @@ private function getTemplateVarVouchers(Cart $cart)
...
     }
 
+    /**
+     * @param array $cartVoucher
+     *
+     * @return bool
+     */
+    private function cartVoucherHasFreeShipping($cartVoucher)
+    {
+        return isset($cartVoucher['free_shipping']) && $cartVoucher['free_shipping'];
```
Comment (PierreRambaud) [inline src/Adapter/Presenter/Cart/CartPresenter.php]:
```suggestion
        return !empty($cartVoucher['free_shipping']);
```

File: src/Adapter/Presenter/Cart/CartPresenter.php [hunk 5]
```diff
@@ -585,6 +608,44 @@ private function getTemplateVarVouchers(Cart $cart)
...
+    }
+
+    /**
+     * @param array $cartVoucher
+     *
+     * @return bool
+     */
+    private function cartVoucherHasPercentReduction($cartVoucher)
+    {
+        if (
```
Comment (PierreRambaud) [inline src/Adapter/Presenter/Cart/CartPresenter.php]:
```suggestion
        return isset($cartVoucher['reduction_percent'])
            && $cartVoucher['reduction_percent'] > 0
            && $cartVoucher['reduction_amount'] == '0.00';
```

File: src/Adapter/Presenter/Cart/CartPresenter.php (hunk 4 above)
Comment (matthieu-rolland) [reply src/Adapter/Presenter/Cart/CartPresenter.php]: done !

File: src/Adapter/Presenter/Cart/CartPresenter.php (hunk 5 above)
Comment (matthieu-rolland) [reply src/Adapter/Presenter/Cart/CartPresenter.php]: done !

File: src/Adapter/Presenter/Cart/CartPresenter.php [hunk 6]
```diff
@@ -585,6 +608,46 @@ private function getTemplateVarVouchers(Cart $cart)
...
+     * @param array $cartVoucher
+     *
+     * @return bool
+     */
+    private function cartVoucherHasPercentReduction($cartVoucher)
+    {
+        return isset($cartVoucher['reduction_percent'])
+            && $cartVoucher['reduction_percent'] > 0
+            && $cartVoucher['reduction_amount'] == '0.00';
+            isset($cartVoucher['reduction_percent'])
```
Comment (PierreRambaud) [inline src/Adapter/Presenter/Cart/CartPresenter.php]: You accept my change without reading it :sweat_smile:

Comment (matthieu-rolland) [reply src/Adapter/Presenter/Cart/CartPresenter.php]: damn I read too fast, I thought you only removed the if and put the condition directly in the return...

Comment (matthieu-rolland) [reply src/Adapter/Presenter/Cart/CartPresenter.php]: Fixed .

**Review by PierreRambaud — APPROVED** [review]

**Comment by matks** [comment]
Thank you @matthieu-rolland

**Comment by alfonsolr09** [comment]
@matthieu-rolland  @PierreRambaud @Robin-Fischer-PS @LouiseBonnard 
I think it would be better if the "discount (s)" section were eliminated
and added where the coupon is applied, see the images


my idea
before: https://imgur.com/8h5GsWZ
after: https://imgur.com/a/lSLsMtB

**Comment by LouiseBonnard** [comment]
@alfonsolr09, this issue/pull request is closed/merged, please create a feature request [here](https://github.com/PrestaShop/PrestaShop/issues/new/choose) so that your idea can be studied. :-)

**Pull request merged**